library/std/src/io/mod.rs

```diff
@@ -3064,6 +3072,53 @@ impl<T> SizeHint for Take<T> {
     }
 }
 
+#[stable(feature = "seek_io_take", since = "CURRENT_RUSTC_VERSION")]
+impl<T: Seek> Seek for Take<T> {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        let offset_from_start = match pos {
+            SeekFrom::Start(offset) => offset,
+            SeekFrom::End(offset) => {
+                if offset > 0 {
```

Comment:
what's with the funny python-indexing-style logic? Rust generally doesn't do that. it doesn't match what's documented for [`Seek::seek`'s errors](https://doc.rust-lang.org/std/io/trait.Seek.html#errors) or [`SeekFrom`](
https://doc.rust-lang.org/std/io/enum.SeekFrom.html).
it also doesn't match how seeking works on [unix](https://manpages.debian.org/bookworm/manpages-dev/lseek.2.en.html#EINVAL)

Reply:
> what's with the funny python-indexing-style logic?

ok, maybe your logic doesn't do that but it is written in a very confusing way with `unsigned_abs` everywhere.

Reply:
a hopefully easier to understand implementation:
(I also implemented some other `Seek` methods and implemented the range-limited seeking as suggested in https://github.com/rust-lang/libs-team/issues/555#issuecomment-2699703221)

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=fa54e42bd394844eb9bf9eb5bcb87ecb

Reply:
The usage of `unsigned_abs` was to deal with `u64`, `i64` and the fact that I need to seek the `inner` from `Current`. I'm ok with your implementation though. I have one questions: the implementation you proposes makes use of the unstable features `unsigned_signed_diff` and `seek_stream_len`. Is it fine to use unstable features in the rust library? Should I just add the two unstable features to the crate attributes?

Reply:
afaik it's generally fine to use unstable library features in the rust standard library, those are implemented in the same git repository, so if they need to change or are removed, the code using them can be updated at the same time.

Reply:
@programmerjake I pushed your implementation with a132fcb. I also removed the tests that now fail and added the needed unstable feature to `library/std/src/lib.rs`.

Reply:
nice! you'll probably want to add some tests that out-of-bounds seeks actually error.

Reply:
Good point. I added five new tests with 6ee1a13:

1. `take_seek_out_of_bounds_start` checks `Start` with offset bigger than `len`
2. `take_seek_out_of_bounds_end_forward` checks  `End` with positive non-zero offsewt
3. `take_seek_out_of_bounds_end_before_start` checks `End` with offset smaller than `-len`
4. `take_seek_out_of_bounds_current_before_start` checks `Current` with offset that pushes position before zero
5. `take_seek_out_of_bounds_current_after_end` checks `Current` with offset that pushes position after `len`